issue #1998 - raise friendly error when checkpoint too close from tip

Elnaril opened the pull request:
### What was wrong?
Issue #1998 

Trinity crashes with a HeaderNotFound exception when launched with `--sync-from-checkpoint eth://block/byhash/` with the checkpoint too close from the chain's tip.

The issue concerns the beam sync when the checkpoint is not strictly more than `MAX_BACKFILL_BLOCK_BODIES_AT_ONCE` (100 000) blocks away from the tip.

The reason is because the beam sync backfills block by chunks of  `MAX_BACKFILL_BLOCK_BODIES_AT_ONCE` and looks first into the DB for their headers.

So, if the specified checkpoint is not before that, some headers will be missing which will trigger the crash.

### How was it fixed?
~~As soon as we get the block number of the chain's tip, we check the checkpoint is strictly before the tip - `MAX_BACKFILL_BLOCK_BODIES_AT_ONCE`.
If not, we end the application with a clear error message.~~

We ensure the block backfiller `BodyChainGapSyncer` fills the most recent actionable gap if one is available, where an actionable gap is as block gap for which:
 - the headers are already in DB
 - the header preceding this block gap is also in DB
 - the size of the gap is <= `MAX_BACKFILL_BLOCK_BODIES_AT_ONCE`

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Add a specific unit test.

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [x] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://assets.rebelmouse.io/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpbWFnZSI6Imh0dHBzOi8vYXNzZXRzLnJibC5tcy8xNjUyNzI2Ni9vcmlnaW4uanBnIiwiZXhwaXJlc19hdCI6MTYxNjgzNDAzMH0.89rYdzbSeojy65TYo1CLIDTYdCy5NIvtkt46YGQQEOk/img.jpg?width=980)

Elnaril commented:
Hello @cburgdorf 
The way I fixed this issue makes test_sync.py to fail.
Before I start digging into this test, could you please share your thoughts on my overall approach, when you have a few minutes ?

Elnaril commented:
Also, I may need to add a check on `self._enable_backfill` along with:
 
    if self._is_checkpoint_too_close_from_tip(final_headers[0].block_number):

or move the whole block inside:

        if self._enable_backfill:
            # There's no chance to introduce new gaps after this point. Therefore we can run this
            # until it has filled all gaps and let it finish.
            self.manager.run_child_service(self._header_backfill)

            # In contrast, block gap fill needs to run indefinitely because of beam sync pivoting.
            self.manager.run_daemon_child_service(self._block_backfill)

            # Now we can check the lag (presumably ~0) and start backfill
            self.manager.run_daemon_task(self._monitor_historical_backfill)

cburgdorf commented on trinity/sync/beam/chain.py:
```diff
@@ -273,6 +285,18 @@ async def run(self) -> None:
         await self.manager.wait_finished()
 
+    def _is_checkpoint_too_close_from_tip(self, tip_block_number: BlockNumber) -> bool:
+        if type(self._launch_strategy) is FromCheckpointLaunchStrategy:
```
I think instead of checking the underlying type of the `_launch_strategy` we may be better of to put this check directly into `FromCheckpointLaunchStrategy`.  It can then raise an exception (maybe `ValidationError`?) that we would have to catch wherever `self._launch_strategy.fulfill_prerequisites()` is called. This will also ensure that the mechanism works not only with `--sync-mode beam` (the default)  but also with `--sync-mode header`.

Elnaril commented:
@cburgdorf : Thank you very much for your feedbacks ! :-) I'll look into them asap.

Elnaril commented on trinity/sync/beam/chain.py:
```diff
@@ -273,6 +285,18 @@ async def run(self) -> None:
+        if type(self._launch_strategy) is FromCheckpointLaunchStrategy:
+            launch_strategy = cast(FromCheckpointLaunchStrategy, self._launch_strategy)
+            if launch_strategy.min_block_number > (
+                tip_block_number - MAX_BACKFILL_BLOCK_BODIES_AT_ONCE - 1
```
@cburgdorf : I may be missing something, but from what I see in my debug sessions, the `HeaderNotFound` exception is raised by the last line of the below code:


    async def _setup_for_next_gap(self) -> None:
        gap_start, gap_end = self._get_next_gap()
        fill_start = BlockNumber(max(
            gap_start,
            gap_end - MAX_BACKFILL_BLOCK_BODIES_AT_ONCE,
        ))
        start_num = BlockNumber(fill_start - 1)
       _starting_tip = await self._db.coro_get_canonical_block_header_by_number(start_num)

where `start_num` depends directly on `MAX_BACKFILL_BLOCK_BODIES_AT_ONCE`.
Or is there something I didn't well identify in this process ?

Elnaril commented:
@cburgdorf : 
Here is my understanding of the issue. Please correct me where I'm wrong ! :)
- The `HeaderNotFound` is raised by `BodyChainGapSyncer._setup_for_next_gap()` when it tries to get the header of the block number #(Tip - MAX_BACKFILL_BLOCK_BODIES_AT_ONCE -1)
 - `FromCheckpointLaunchStrategy.fulfill_prerequisites()` does not raise it and does not know yet the tip number, which seems to be saved in DB only after the following lines from `BeamSyncer.run()`:

        async with background_asyncio_service(self._header_persister) as manager:
            await manager.wait_finished()
- Still according to my understanding, `--sync-mode header` would not backfill block bodies nor have a mechanism which looks for a header in DB before the checkpoint. That would explain why I don't reproduce the issue in this mode.

cburgdorf commented:
@Elnaril I think there's some misunderstanding about the involvement of block bodies here. Let's see if I can clear it up :)

**SPOILER:** There's a plot twist at 7. but I didn't want to throw away everything I had written before :joy: 

1. Block bodies aren't a concern here at all. All we need to care about is that the block number (which can be read from the block header, no block body needed) that was chosen isn't too close to the tip of the chain. This check is important for every sync strategy that the `--sync-from` argument can be used with which includes `--sync-mode header`.

2. Because of the former, the constant that we need to care about is `MAX_SKELETON_REORG_DEPTH` not `MAX_BACKFILL_BLOCK_BODIES_AT_ONCE`

3. When syncing kicks off, it launches according to a given `SyncLaunchStrategyAPI` which has a method `fulfill_prerequisites` that can perform any kind of work that needs to be done *before* syncing can start. It's like a pre-hook.

https://github.com/ethereum/trinity/blob/aafc20c63716f018b0a628e3b632471b77a7c1dc/trinity/sync/common/strategies.py#L48-L56

4. There are cases where the client might fail to fulfill the prerequisites in which case an exception is caught and the client is stopped with an error message.

Header sync:
https://github.com/ethereum/trinity/blob/aafc20c63716f018b0a628e3b632471b77a7c1dc/trinity/sync/header/chain.py#L67-L75

Beam sync:
https://github.com/ethereum/trinity/blob/aafc20c63716f018b0a628e3b632471b77a7c1dc/trinity/sync/beam/chain.py#L210-L218

There's some duplication :point_up: with catching these failures for the different sync modes but I think that shouldn't be a big concern here.

5.There are multiple implementations of `SyncLaunchStrategyAPI` but the one we care about here is `FromCheckpointLaunchStrategy` which launches sync from a given checkpoint.


6. Currently, the only case where `FromCheckpointLaunchStrategy` would raise an exception is when we fail to get the chosen header from any peer within a finite set of attempts

https://github.com/ethereum/trinity/blob/aafc20c63716f018b0a628e3b632471b77a7c1dc/trinity/sync/common/strategies.py#L242-L244

7. Now that I wrote all this, I realize that there is already a check that checks if we are too close to the tip but I think that check might be based on a different (It bases of `FULL_BLOCKS_NEEDED_TO_START_BEAM`)

https://github.com/ethereum/trinity/blob/aafc20c63716f018b0a628e3b632471b77a7c1dc/trinity/sync/common/strategies.py#L196-L208

Looking back at the original ticket #1998 I noticed that the crash originates from a different place than what I had in mind when I wrote all of the above.

```
File "/home/cburgdorf/Documents/hacking/ef/trinity/trinity/sync/beam/chain.py", line 531, in run
    await self._setup_for_next_gap()
  File "/home/cburgdorf/Documents/hacking/ef/trinity/trinity/sync/beam/chain.py", line 452, in _setup_for_next_gap
    _starting_tip = await self._db.coro_get_canonical_block_header_by_number(start_num)
  File "/home/cburgdorf/Documents/hacking/ef/trinity/trinity/_utils/async_dispatch.py", line 22, in wrapper
    return await loop.run_in_executor(
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/cburgdorf/Documents/hacking/ef/trinity/venv/lib/python3.8/site-packages/eth/db/header.py", line 121, in get_canonical_block_header_by_number
    return self._get_canonical_block_header_by_number(self.db, block_number)
  File "/home/cburgdorf/Documents/hacking/ef/trinity/venv/lib/python3.8/site-packages/eth/db/header.py", line 129, in _get_canonical_block_header_by_number
    canonical_block_hash = cls._get_canonical_block_hash(db, block_number)
  File "/home/cburgdorf/Documents/hacking/ef/trinity/venv/lib/python3.8/site-packages/eth/db/header.py", line 114, in _get_canonical_block_hash
    raise HeaderNotFound(
eth.exceptions.HeaderNotFound: No canonical header for block number #3196818
```

It originates from `_setup_for_next_gap` here

https://github.com/ethereum/trinity/blob/aafc20c63716f018b0a628e3b632471b77a7c1dc/trinity/sync/beam/chain.py#L453

So *maybe* the real fix isn't demanding a bigger distance to the tip but figuring out why the gap syncer (the background sync that fills the gaps that occur when using checkpointing) gets confused about the gaps and crashes.

If I look at the original issue it suggests that one **can** reproduce this issue by using a checkpoint to close to the tip and fix the issue by choosing a checkpoint more far away from the tip but having spend the time to write this up I'm now not convinced if that is the actual proper fix or if there's a different problem to be solved. All I can say is that it probably needs more investigation...

Elnaril commented:
@cburgdorf : Thank you very much for your feedback, I really appreciate it ! :-)

Elnaril commented:
Hello @cburgdorf 
I believe `_get_next_gap()` should return actual gaps when there is a checkpoint, meaning gaps which are only after this checkpoint. 
In order to accomplish that, maybe we could have a specialized gap syncer:
`class FromCheckpointBodyChainGapSyncer(BodyChainGapSyncer):`
This gap syncer would know about the checkpoint and so would be able to filter the gaps.
I committed an illustration of this idea, so we can have a starting point of discussion in case this is of any interest.

cburgdorf commented:
@Elnaril First of all, thanks for taking the time to investigate that further. My head is currently [elsewhere](https://github.com/ethereum/fe) which is making it a bit harder for me to keep up with this. I haven't yet fully understood what the problem is so I have a few questions:

1. Is this problem only caused by the backfilling of blocks or also by backfilling of headers (which are two distinct operations)?

2. Can you describe step for step the scenario that leads to the crash?

3. Even better, can you come up with a test to reproduce the issue? We have some tests that cover the gap filler so if we can come up with a test that clearly demonstrates the problem that would be great!. Here's a good starting point.

https://github.com/ethereum/trinity/blob/07e259b9fcaa89dc008836fbd9bb180b56c715b8/tests/core/p2p-proto/test_sync.py#L832

Sorry, if that isn't much of support right now but the issue isn't at the top of my todo list right now so I'm trying to limit my review time while at the same time trying to provide some pointers to help you move it forward :sweat_smile:

Elnaril commented:
@cburgdorf Thank you very much for your time ! :-) I will definitively try to make good use of your pointers ! :smiley:

Elnaril commented:
Hello @cburgdorf 

1/ As we saw previously, the error is raised at the beginning of the blocks backfilling, when the blocks syncer (`BodyChainGapSyncer`) tries to get the header of the block number `#(Tip - MAX_BACKFILL_BLOCK_BODIES_AT_ONCE -1)` from the DB (in the method `_setup_for_next_gap()`). This header is not in DB, hence the `HeaderNotFound` exception.

There is no error with the  backfilling of headers. The reason is that the headers syncer starts to backfill all headers from block 1 by intervals of 100000):

_INFO  2020-11-08 15:44:22,011  HeaderChainGapSyncer  Initializing gap-fill header-sync; filling gap: (**1, 100001**)_

and the headers backfilling process checks (in `fulfill_prerequisites()`) if the header of block 0 is in DB, which is the case for some reason:
`self._get_canonical_block_header_by_number(self.db, 0)` finds a header, but `self._get_canonical_block_header_by_number(self.db, 1)` and `self._get_canonical_block_header_by_number(self.db, #Tip -100000 + 1)` don't.

So no error for backfilling of headers, but I wonder if backfilling of headers from block 1 is expected for a `FromCheckpointLaunchStrategy` ?

2/ To reproduce the crash, we need to launch the beam syncer with `--sync-from-checkpoint` and, as checkpoint, a block whose number is not strictly more than MAX_BACKFILL_BLOCK_BODIES_AT_ONCE (100 000) blocks away from the tip.
So, for Goerli with a tip block number =  3718888
 - checkpoint 3630000 => crash with HeaderNotFound
 - checkpoint 3600000 => no crash

3/ I've added 2 tests in /trinity/tests/core/p2p-proto/test_sync.py to demonstrate the issue:
 - `test_block_gapfill_from_checkpoint_syncer_1()` which simulates the backfilling of bodies with BodyChainGapSyncer and fails (`eth.exceptions.HeaderNotFound: No canonical header for block number #898` in the logs of _py37-eth1-core_).
 - `test_block_gapfill_from_checkpoint_syncer_2()` which simulates the backfilling of bodies with FromCheckpointBodyChainGapSyncer (the class I added previously to illustrate a fix) and succeeds.

Both tests create a new chain with headers 970 to 1000, and force the `syncer._max_backfill_block_bodies_at_once` to 100 (I added this field (which exists for `SequentialHeaderChainGapSyncer`(self._max_backfill_header_at_once)) to allow the test), then start the body backfilling. The tests are meant to succeed if the blocks 970 to 1000 are imported (but not block < 970).
I struggled a bit to make a test chain with only the headers and had to add a block otherwise no block gaps were found ...

4/ About the fix: In my understanding, the body syncer should backfill only blocks after the checkpoint, so it should not get gaps before that checkpoint. FromCheckpointBodyChainGapSyncer is made aware of the checkpoint number and therefore can filter those wrong gaps.

cburgdorf commented:
Hi @Elnaril thanks for your work and detailed writeup. If I understand the issue correctly, it could be that the bug might have been introduced in https://github.com/ethereum/trinity/commit/ae95f78d8fdfb01158df1711c9e3081e5cd5c559 (cc @carver) which tries to improve beam sync by prioritizing backfilling of more recent blocks over older ones. I'm speculating but it would probably be worth checking a previous commit to see if the bug existed before that change.

>In my understanding, the body syncer should backfill only blocks after the checkpoint

I think the general assumption about these backfill tasks is that you should be able to start Trinity with any form of chain fragmentation (e.g. maybe you started it with several different checkpoints in the past and hence has lots of gaps) and if you just keep it running long enough it will eventually fill all gaps so that you arrive at a chain that is whole from genesis all the way to the tip.

Given this assumption, I don't think we can filter out gaps entirely.

To me, it seems that there is something wrong with **how we recognize** *chain gaps*.

Let's be clear about the two different forms of gaps that we track. There are gaps in the **chain of headers** that we can read via the `get_header_chain_gaps` API.

https://github.com/ethereum/trinity/blob/4bcd5dc3dd148039cf0de5a19b2472240d0e4693/trinity/sync/beam/chain.py#L480

If we don't have the header, we are also missing the block but that is not what we recognize as a gap in the **chain of blocks**. We do only count gaps in the chain of blocks for blocks where we **do have the header**. This is what we can read with the `get_chain_gaps` API.

https://github.com/ethereum/trinity/blob/4bcd5dc3dd148039cf0de5a19b2472240d0e4693/trinity/sync/beam/chain.py#L476

If there are no gaps in the chain of blocks, we check if there is room to catch up from our highest written block to our highest consecutive written header.

https://github.com/ethereum/trinity/blob/4bcd5dc3dd148039cf0de5a19b2472240d0e4693/trinity/sync/beam/chain.py#L477-L493

So, when we hit this code, our assumption is that we can always lookup the immediate header before the gap start.

https://github.com/ethereum/trinity/blob/4bcd5dc3dd148039cf0de5a19b2472240d0e4693/trinity/sync/beam/chain.py#L448-L452

But for some reason, that assumption doesn't hold and I haven't fully understood why. It seems to me as if `gap_start` might already be a checkpoint header and we are trying to reach one header backwards from here which then fails. But I don't see how we would end up with this scenario.

Anyway, I think that your current solution is trying to work around the issue in a way that will simply leave gaps unfilled which I don't think is acceptable (How are we supposed to fill them eventually?).

I could be wrong (and so I'll summon @carver for another pair of :eyes: because he's so much smarter than I :sweat_smile: ) but I want to believe the solution is hidden in `def _get_next_gap` and that this function may currently not always return what we think it does.

Elnaril commented:
Hi @cburgdorf and thank you very much for your time and feedbacks ! :-)

>I think the general assumption about these backfill tasks is that you should be able to start Trinity with any form of chain fragmentation (e.g. maybe you started it with several different checkpoints in the past and hence has lots of gaps) and if you just keep it running long enough it will eventually fill all gaps so that you arrive at a chain that is whole from genesis all the way to the tip.

Thank you for this clarification, it makes totally sense and indeed my solution was not heading this way at all !!
My understanding of the purposes of launching from a checkpoint was partially wrong. :sweat_smile:

I'll have a look at https://github.com/ethereum/trinity/commit/ae95f78d8fdfb01158df1711c9e3081e5cd5c559 and how it worked before this commit and based on all your very helpful clarifications, I'll try to figure out a better way to solve the issue.

Elnaril commented:
Hello @cburgdorf ,
Here is where my investigation stands at the moment. Please let me know if you disagree with anything.

1/ From what I can see after a few tests, the commit https://github.com/ethereum/trinity/commit/ae95f78d8fdfb01158df1711c9e3081e5cd5c559 , _per se_, did not cause the issue : before this commit, there would be a crash because of the missing header 1 (header 0 is in DB).
2/ I had a look into [test_eth1_chaindb.py/test_block_gap_tracking()](https://github.com/ethereum/py-evm/blob/master/tests/database/test_eth1_chaindb.py) and it seems to me there is a unit test corresponding to our situation:

https://github.com/ethereum/py-evm/blob/5b177029d4d08454fa097bcd451c89b1d4a92e89/tests/database/test_eth1_chaindb.py#L173-L184

In this test, header 4 and 5 are missing and as soon as block 7 is added, block 4, 5 and 6 are found missing by `get_chain_gaps()`
That seems to be very similar to our case where block 0 and some blocks near the tip are present but not all headers, and `get_chain_gaps()` finds the all blocks in between missing.

Elnaril commented:
So, if my understanding is correct, would that mean `_get_next_gap()` should be adapted and use also `get_header_chain_gaps()` to compute the right block gaps ?

cburgdorf commented:
>So, if my understanding is correct, would that mean `_get_next_gap()` should be adapted and use also `get_header_chain_gaps()` to compute the right block gaps ?

It does use `get_header_chain_gaps()` already.

https://github.com/ethereum/trinity/blob/f33bf1260be68539d5d706c8c0e62b1f79fe6b4a/trinity/sync/beam/chain.py#L488-L490

If there are no actual *gaps* in the chain of blocks (which I assume to be the case, especially if this error happens with a fresh database), then it calls `get_header_chain_gaps()` to find out if there is room to catch up to the *highest consecutive written header*. Where *highest consecutive written header* means the header with the highest number that we already have which is perfectly connected without any gaps all the way down to genesis.

So, lets assume we start Trinity and the DB currently looks like this:

Headers: 0 - 10000 (no gaps)
Blocks: None

We start Trinity with Checkpoint header 20000.

The check should say:

- We don't have any block gaps
- Oh but we have a perfect header chain up to header 10000
- Let's fetch blocks 0 to 10000 then.

This leads me to the question: Can we confirm that this error does really happen with a fresh database (you can delete the db or use `--trinity-tmp-root-dir` to start it from a fresh temporary directory)

Because with a fresh database I assume it *should* play out like this:

Headers: 0 (Genesis Header is hard wired)
Blocks: None

We start Trinity with Checkpoint header 20000.

The check should say:

- We don't have any block gaps
- Our highest header is the genesis header
- We can only fetch the genesis block

If this error happens with a fresh database and tries to fetch blocks far in the future, then I assume there must be some error in `_get_next_gap` even if it isn't obvious to me from starring at the code.

If this error doesn't happen with a fresh database, then it is likely more subtle.

Elnaril commented:
Hello @cburgdorf and thank you again for your time, I'm sure we're getting close to figure out the root cause of this issue ! :-)

Firstly, I do all my debug sessions on Goerli after deleting the folders I specified in `--trinity-root-dir`. So, yes the issue does happen with a fresh database.

> It does use get_header_chain_gaps() already.
https://github.com/ethereum/trinity/blob/f33bf1260be68539d5d706c8c0e62b1f79fe6b4a/trinity/sync/beam/chain.py#L488-L490
> If there are no actual gaps in the chain of blocks (which I assume to be the case, especially if this error happens with a fresh database), then it calls get_header_chain_gaps() to find out if there is room to catch up to the highest consecutive written header. Where highest consecutive written header means the header with the highest number that we already have which is perfectly connected without any gaps all the way down to genesis.

Yes, I totally agree with the above. The thing is that the crash happens before we get a chance to go into that block.

Here is what exactly happens when I launch it on Goerli with a fresh DB and a checkpoint at block number 3690000 (for the record, according to goerli.etherscan.io, the last block was ~3781442 at the moment I launched Trinity and ~3781501 when `BodyChainGapSyncer` started to run):

![image](https://user-images.githubusercontent.com/57373038/99679547-f5e58f00-2a73-11eb-988f-569244d82ca6.png)
I added lines 479 and 480 to help understanding the situation.
So when `BodyChainGapSyncer` runs, the very first time we enter `_get_next_gap(self)` we have the following (breakpoint on line 481):
![image](https://user-images.githubusercontent.com/57373038/99680075-93d95980-2a74-11eb-9148-5cbff1c63502.png)
As you can see:
 - `self._db.get_chain_gaps()` gets a blocks gap between blocks number 1 and 3781494
 - `self._db.get_header_chain_gaps()` gets a headers gap between 1 and 3689999
 - `len(gaps) == 1` so we don't reach line 484. Instead the method returns `gaps[-1]`
 - Genesis block does exist (and not only its header)

Then we come back to `_setup_for_next_gap(self)`:
![image](https://user-images.githubusercontent.com/57373038/99682826-938e8d80-2a77-11eb-925b-821e3c127fa0.png)

![image](https://user-images.githubusercontent.com/57373038/99682541-3a265e80-2a77-11eb-98f8-f0a172524b30.png)

where we look for header number 3681493 in DB which fails since it's in the headers gap (1, 3689999) and Trinity crash with the exception :
![image](https://user-images.githubusercontent.com/57373038/99683261-0ef03f00-2a78-11eb-8b50-d01c33bcb470.png)


So, to summarize the issue:
1/ we start with a fresh DB, no blocks, no headers
2/ we launch Trinity from checkpoint 3690000
3/ at the moment the backfilling of blocks starts, we have in DB:
  - the headers after (and including) 3690000
  - the genesis block (so header + body)
  - the block 3781495 (and after):
![image](https://user-images.githubusercontent.com/57373038/99684062-ec125a80-2a78-11eb-937d-a6f058fb1b35.png)

4/ then, `self._db.get_chain_gaps()` report a blocks gap between 1 and 3781494
5/ we start to backfill from 3781494 - 100000 = 3681494
6/ for that we look for the previous header (3681494) in DB which does not exist
7/ crash

To conclude, it seems to me that the following assumption is violated, since the genesis block exists and not only its header:

> Because with a fresh database I assume it should play out like this:
> 
> Headers: 0 (Genesis Header is hard wired)
> Blocks: None

If we had only the genesis header, and not its body, no block gap would be retrieved before the checkpoint.

(sorry, for this long post ...)

cburgdorf commented:
Great analysis @Elnaril !,

>4/ then, `self._db.get_chain_gaps()` report a blocks gap between `1` and `3781494`

Ok, this is interesting because I would have assumed that for this to be a valid block gap it means we need to have block `3781494`. But even if we ignore that, [the test that you linked](https://github.com/ethereum/py-evm/blob/5b177029d4d08454fa097bcd451c89b1d4a92e89/tests/database/test_eth1_chaindb.py#L173-L184) clearly shows that we can have block gaps for ranges that both blocks **and** headers are missing.

And clearly `_get_next_gap` does not account for that. It just blindly returns the block gap without verifying that it is actionable (meaning, we have the headers for that gap).

https://github.com/ethereum/trinity/blob/f33bf1260be68539d5d706c8c0e62b1f79fe6b4a/trinity/sync/beam/chain.py#L506-L507

So, I'm ignoring the questions **why** we see this specific block gap being reported because the tests actually demonstrate that we need to account for cases like that. So, it seems to me that what we need is an extra check to find out if the gap is actionable. 

I *think* what we need goes roughly like this:

```
    def _get_next_gap(self) -> BlockRange:
        gaps, future_tip_block = self._db.get_chain_gaps()
        try:
            actionable_gap = get_topmost_actionable_gap(gaps)
        except NoActionableGap:
            # We do not have gaps in the chain of blocks but we may still have a gap from the last
            # block up until the highest consecutive written header.
            header_gaps, future_tip_header = self._db.get_header_chain_gaps()
            if len(header_gaps) > 0:
                # The header chain has gaps, find out the lowest missing header
                lowest_missing_header, _ = header_gaps[0]
            else:
                # It doesn't have gaps, so the future_tip_header is the lowest missing header
                lowest_missing_header = future_tip_header

            highest_consecutive_header = lowest_missing_header - 1
            if highest_consecutive_header >= future_tip_block:
                # The header before the lowest missing header is the highest consecutive header
                # that exists in the db and it is higher than the future tip block. That's a gap
                # we can try to close.
                return future_tip_block, BlockNumber(highest_consecutive_header)
            else:
                raise ValidationError("No gaps in the chain of blocks")
        else:
            return actionable_gap
```

Where `get_topmost_actionable_gap` would loop through the gaps and finds the most recent gap which is actionable, meaning where we do have the previous header available.

>If we had only the genesis header, and not its body, no block gap would be retrieved before the checkpoint.

I'm not really sure if that is the case but it seems to me the tests are prove enough that gaps in the chain of blocks can be reported for ranges where both blocks and headers are missing. Hence my proposal above to address that.

>(sorry, for this long post ...)

That was a really helpful analysis. Thanks for investing so much time into getting to the bottom of this case!

Elnaril commented:
Thank you @cburgdorf for your always very helpful guidances ! :-)

I created the exception class `NoActionableGap` in trinity/sync/beam/chain.py , but maybe it would be better to define it elsewhere ? Maybe in trinity/exceptions.py ? Also, should it inherit from `BaseTrinityError` ?

Moreover, I created the `get_topmost_actionable_gap()` method.

Because of line 479 in the following code:
https://github.com/ethereum/trinity/blob/f33bf1260be68539d5d706c8c0e62b1f79fe6b4a/trinity/sync/beam/chain.py#L475-L482
we would need to have the headers for all the blocks in the block gap in order to be actionable, and not just the the header of the block preceding the gap. So this method uses the header and block gaps to return the most recent gap of blocks of max size = `MAX_BACKFILL_BLOCK_BODIES_AT_ONCE` for which the headers exist in DB, and also the preceding header.

This method leaves room for optimization (we could identify a smaller actionable gap i/o raising immediately a NoActionableGap), but I believe it is a good balance between complexity/maintenance and efficiency for a backfilling process, at least for now and in particular regarding the following.

So, speaking of optimization, I noticed the backfilling of the headers starts from block 1:
![image](https://user-images.githubusercontent.com/57373038/100109127-6ec55c00-2e63-11eb-81be-8f52aa3babf7.png)

In our case, where we start from a checkpoint and especially with a fresh DB, that will make the block backfilling wait for all the headers to be imported up to the checkpoint, which seems to defeat the point of "preferring recency" (cf ae95f78d8fdfb01158df1711c9e3081e5cd5c559).

Back to my commits: I also added a pytest method `test_block_gapfill_from_checkpoint_syncer()` in tests/core/p2p-proto/test_sync.py to illustrate the new method and test the bug is fixed.

Looking at the CI checks, it looks like I will need to rebase ...

> That was a really helpful analysis. Thanks for investing so much time into getting to the bottom of this case!

I am really glad if my PRs are helpful ! :-) I do hope they are worth the time you guys spend to review them !

cburgdorf commented:
>but maybe it would be better to define it elsewhere ?

I think it's ok to live there locally

>Also, should it inherit from `BaseTrinityError` ?

Yes, it should.

>we would need to have the headers for all the blocks in the block gap in order to be actionable, and not just the the header of the block preceding the gap

I'm thinking that the body gap syncer will *include* the headers which means we only need to check that the previous header exists because everything that follows will be fetched in tandem header/block. I could be wrong but at a first glance this is how I *think* it works.

I'm not sure if `get_topmost_actionable_gap` needs to be concerned with `MAX_BACKFILL_BLOCK_BODIES_AT_ONCE`. I would think that we would just check if it is actionable (for which I continue to think we could just check for the parent header but I might be wrong. Did you try it out?) but don't actually care for the *size* of the gap because restricting the size of the gap is already done by the calling code in `_setup_for_next_gap`

https://github.com/ethereum/trinity/blob/f33bf1260be68539d5d706c8c0e62b1f79fe6b4a/trinity/sync/beam/chain.py#L458-L461

>In our case, where we start from a checkpoint and especially with a fresh DB, that will make the block backfilling wait for all the headers to be imported up to the checkpoint, which seems to defeat the point of "preferring recency" 

I think if you start with a fresh db and a checkpoint then it will play out somehow like this:

1. Start with Checkpoint `1000`
2. The checkpoint is already a safety distance away from the tip so we first start syncing headers only
3. At some point we are close enough to the tip (say we are at `3000` now).
4. We start executing blocks so everything after 3000 we do have headers + blocks (and even executed them, yay beam sync!)
5. But now we can catch up blocks 1000 to 3000 because we only have the headers for those yet
6. Last but not least, we start filling the gap from `1` to `1000`.

>Looking at the CI checks, it looks like I will need to rebase

Yep, seems so.

>I am really glad if my PRs are helpful ! :-) I do hope they are worth the time you guys spend to review them !

No, you do provide a useful work and I'm happy you have taken on this challenge which I initially thought to be much simpler but when I filed the issue I didn't have all the context that you provided with your ongoing work.

Elnaril commented:
Hello @cburgdorf , and thanks for your feedback. I may miss something, but I'll try to clarify my points.

> for which I continue to think we could just check for the parent header but I might be wrong. Did you try it out?

Actually I faced it during one of my previous investigations. The reason I figured out at this moment is that `DatabaseBlockRangeHeaderSyncer(self._db, (fill_start, gap_end,)),` is initialized by getting the header number `gap_end` from the DB (line 564 below) and then will yield all headers in `range(fill_start, gap_end +1)` from DB as well (lines 575 to 581 below).
=> I made the assumption at this moment that all the headers of the gap had to be in DB before the gap could be filled with blocks, but didn't push the investigation further.

https://github.com/ethereum/trinity/blob/f33bf1260be68539d5d706c8c0e62b1f79fe6b4a/trinity/sync/common/headers.py#L553-L581

So, to demonstrate it, I wrote an ad hoc test:

**PART I**
I built a chain with:
 - Header gaps: (1, 897), (899, 969)  => header 898 is in DB
 - Body gap: (1, 999)
 - and I force `get_topmost_actionable_gap()` to return (899, 950) which is a gap without header in DB. Only header 898 is in DB.

**/!\ Now that I have done this part, I wonder if that is a realistic production case as we may have always some headers preceding the end of the body gap ? Not sure though, but you may want to skip this part and go to PART II directly ...**

Here is the test code:
![image](https://user-images.githubusercontent.com/57373038/100222431-f8336780-2f11-11eb-9a30-a81a51c84167.png)

So, at breakpoint line 783, we have the following:
![image](https://user-images.githubusercontent.com/57373038/100222861-9cb5a980-2f12-11eb-8652-d2ab77446b48.png)

And then, once `BodyChainGapSyncer` runs, in `_setup_for_next_gap()` we have the following at breakpoint line 470:
![image](https://user-images.githubusercontent.com/57373038/100224253-7e50ad80-2f14-11eb-960d-4cc9226fa673.png)
![image](https://user-images.githubusercontent.com/57373038/100224783-2f574800-2f15-11eb-8e6e-40cb186c319f.png)

`DatabaseBlockRangeHeaderSyncer` is then initialized and crashes at line 564 with a `HeaderNotFound` exception:
![image](https://user-images.githubusercontent.com/57373038/100225138-a8ef3600-2f15-11eb-9019-00d31de4b595.png)
![image](https://user-images.githubusercontent.com/57373038/100225379-071c1900-2f16-11eb-9c74-873c64a7141f.png)
![image](https://user-images.githubusercontent.com/57373038/100225231-cb814f00-2f15-11eb-9643-1ac01b0d99e5.png)
![image](https://user-images.githubusercontent.com/57373038/100225754-9f1a0280-2f16-11eb-94c5-166c1af466c9.png)


**PART II**
=> Until now, this test demonstrates we need at least the header of the last block of the actionable gap, in addition of the preceding header, but as I said at the beginning, I'm not sure yet if that is a realistic production case.
So, I changed `get_topmost_actionable_gap()` to return now (899, **999**), for which we don't have the headers 899 to 969 in DB but we still have the precedent header (898).
At breakpoint line 783 we have:
![image](https://user-images.githubusercontent.com/57373038/100226344-84945900-2f17-11eb-8bb5-10be527b4ede.png)
![image](https://user-images.githubusercontent.com/57373038/100226447-adb4e980-2f17-11eb-86fc-12e87c970d8b.png)

This time, all good at `DatabaseBlockRangeHeaderSyncer` initialization:
![image](https://user-images.githubusercontent.com/57373038/100227071-78f56200-2f18-11eb-92c8-5d34c6d5acfb.png)
![image](https://user-images.githubusercontent.com/57373038/100227116-8874ab00-2f18-11eb-89bf-050d7a679f53.png)

But now we enter immediately `new_sync_headers()`:
![image](https://user-images.githubusercontent.com/57373038/100227295-d5588180-2f18-11eb-8fcb-0d523f778823.png)
![image](https://user-images.githubusercontent.com/57373038/100227385-fc16b800-2f18-11eb-8717-7dc68ad92020.png)

And a `HeaderNotFound` exception is raised:
![image](https://user-images.githubusercontent.com/57373038/100244085-76056c00-2f2e-11eb-827b-609c46f5932b.png)


**PART III**
I did a last test. I launched trinity in debug mode on Goerli with a checkpoint on a fresh DB, but with the following modifications:
 - I deactivated the header backfilling in order to avoid any unwanted interferences
 - I deactivated the shortening of the block gap in `_setup_for_next_gap()`: the aim is to have a block gap with the previous header (here, the genesis) being in DB.

![image](https://user-images.githubusercontent.com/57373038/100251378-9cc7a080-2f36-11eb-960f-afc939c192e1.png)
we initialize `DatabaseBlockRangeHeaderSyncer` with fill_start = 1 and gap_end = 3816476:
![image](https://user-images.githubusercontent.com/57373038/100251449-b2d56100-2f36-11eb-819b-4cf591615c67.png)

No issue with the initialization and we get immediately into `new_sync_headers()`
![image](https://user-images.githubusercontent.com/57373038/100251599-d9939780-2f36-11eb-905b-c231915efc2e.png)
![image](https://user-images.githubusercontent.com/57373038/100251669-ed3efe00-2f36-11eb-81e1-664ff948a306.png)
A `HeaderNotFound` exception is raised at the first loop with `block_number` == 1 and Trinity crashes:
![image](https://user-images.githubusercontent.com/57373038/100251797-1069ad80-2f37-11eb-8e1d-807e4aa7b9c9.png)

To conclude, I'm still convinced that either all headers of a block gap have to be in DB in order for the gap to be actionable, or some other substantial changes would be needed.
Regarding the relevance of MAX_BACKFILL_BLOCK_BODIES_AT_ONCE in the computing of actionable gaps, I let that for later when we have clarified the above.

> you do provide a useful work and I'm happy you have taken on this challenge

:blush:

cburgdorf commented:
Ah yes, you are right, the `DatabaseBlockRangeHeaderSyncer` reads the headers out of the DB. I should know better since I wrote that code :see_no_evil: 
But I guess since your solution checks the existence of the headers based on the header chain gaps information (as opposed to trying to make DB reads) checking each header for existence should be ok :ok_hand: 

>Regarding the relevance of MAX_BACKFILL_BLOCK_BODIES_AT_ONCE in the computing of actionable gaps, I let that for later when we have clarified the above.

Ok, let's move on to that one. It's actually not a big deal. I guess I don't see why it would be needed inside `get_topmost_actionable_gap` but maybe it is there to minimize the performance impact of that method (like, why keep checking if the gap is already bigger than what we would end up consuming anyway).

Also, you are much closer to the code right now than I am so I'm :+1: with this solution if it works. So, if you can clean this up and get CI to pass, I'm fine to go with this :)

Elnaril commented:
Cool !! :-)

I actually added the use of MAX_BACKFILL_BLOCK_BODIES_AT_ONCE inside `get_topmost_actionable_gap()` to avoid discarding actionable gaps unnecessarily.
To fix ideas, let's consider the following example:
 - header gap: `(1, 890)`
 - body gap: `(1, 999)`
 - `MAX_BACKFILL_BLOCK_BODIES_AT_ONCE` = 100

##### Case _with_ using `MAX_BACKFILL_BLOCK_BODIES_AT_ONCE`:
 1/ `get_topmost_actionable_gap()` returns the actionable gap (899, 999)
 2/ `_setup_for_next_gap()` keeps (899, 999) and the body backfiller can start immediately in parallel with the header backfiller

##### Case _without_ using `MAX_BACKFILL_BLOCK_BODIES_AT_ONCE`:
 1/ `get_topmost_actionable_gap()` returns no actionable gap, because the only body gap (1, 999) misses headers 1 to 890 in DB. 
 2/ The body backfiller will have to wait for the header backfiller to retrieve the missing headers.

Conclusion: In the case _"without"_, the actionable gap (899, 999) has not been considered even though (1, 999) would have been truncated anyway by `_setup_for_next_gap()` precisely to (899, 999), resulting in an unnecessary wait for the body backfiller.
Please, Let me know if you think that would be better otherwise and I'll update accordingly.

Regarding the CI, despite having rebased, some py37 tests are still failing for reasons which seems, to me, unrelated to my changes. I'll try to dig further, though.

It seems we see the light at the end of the tunnel !! :smile:
Thanks again for all your help ! :-)

Elnaril commented:
Hello @cburgdorf ,

It seems we have a conflict version between `importlib-metadata` and `argcomplete` because of the following:

On one hand:
`argcomplete` v<1.12.2 accepted only version < 3 for `importlib-metada`:
[see argcomplete change here, line 8 of setup.py](https://github.com/kislyuk/argcomplete/compare/v1.12.1...v1.12.2#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R8)

On the other hand:
`importlib-metada` 3 was released on 22nd Nov
https://pypi.org/project/importlib-metadata/#history

Now all py37 checks are good. I guess this issue impacted only py37 checks because `importlib-metada` has been backported from Python 3.8.
That should also help the PR https://github.com/ethereum/trinity/pull/2109 though I'm not sure if that is the best way to solve it ?
And I guess this version fix should be in a separate PR ? Let me know if I should rollback setup.py.

I have now an error on py38-wheel-cli which I had not before. And py37-wheel-cli is all good. So I wonder if we should just retrigger the py38-wheel-cli test ? Is it possible ? Otherwise I'll try to dig into that tomorrow.

cburgdorf commented:
>And I guess this version fix should be in a separate PR

Yes, if you could spin out the `setup.py` fix into its own PR that would be better.

>I have now an error on py38-wheel-cli which I had not before

Yes, I think that error is unrelated. It's still interesting. It looks like a failure in our event bus endpoints cross connection ceremony. I've recently read about that error in the chat but had not seen it before myself.

```
   DEBUG  11-26 21:59:37  async_process_runner.py  b'\x1b[1m\x1b[33m WARNING  2020-11-26 21:59:37,211       EventBusService  Failed to connect network-db to one of discovery: [Errno 11] Resource temporarily unavailable\x1b[0m\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/cytoolz/compatibility.py:2: DeprecationWarning: The toolz.compatibility module is no longer needed in Python 3 and has been deprecated. Please import these utilities directly from the standard library. This module will be removed in a future release.\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  warnings.warn("The toolz.compatibility module is no longer "\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/async_lru.py:96: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  def _wait_closed(wrapped, *, return_exceptions, loop):\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/async_lru.py:191: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  def wrapped(*fn_args, **fn_kwargs):\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/aiohttp/helpers.py:105: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  def noop(*args, **kwargs):  # type: ignore\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'<Manager[AsyncioEventBusService] flags=SRcfe>: task _auto_connect_new_announced_endpoints[daemon=True] exited with error: [Errno 11] Resource temporarily unavailable\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'Traceback (most recent call last):\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/async_service/base.py", line 324, in _run_and_manage_task\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    await task.run()\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/async_service/asyncio.py", line 35, in run\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    await self._async_fn(*self._async_fn_args)\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/trinity/extensibility/event_bus.py", line 103, in _auto_connect_new_announced_endpoints\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    await endpoint.connect_to_endpoints(*endpoints_to_connect_to)\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/lahja/asyncio/endpoint.py", line 529, in connect_to_endpoints\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    await asyncio.gather(\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/lahja/asyncio/endpoint.py", line 537, in _await_connect_to_endpoint\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    conn = await AsyncioConnection.connect_to(endpoint.path)\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/lahja/asyncio/endpoint.py", line 88, in connect_to\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    reader, writer = await asyncio.open_unix_connection(str(path))\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/usr/local/lib/python3.8/asyncio/streams.py", line 111, in open_unix_connection\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    transport, _ = await loop.create_unix_connection(\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "uvloop/loop.pyx", line 2223, in create_unix_connection\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "uvloop/loop.pyx", line 2218, in uvloop.loop.Loop.create_unix_connection\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'BlockingIOError: [Errno 11] Resource temporarily unavailable\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'<bound method AsyncioIsolatedComponent._do_run of <trinity.components.builtin.network_db.component.NetworkDBComponent object at 0x7fc913ddd040>> raised an unexpected exception\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'Traceback (most recent call last):\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/asyncio_run_in_process/_child.py", line 205, in run_process\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    runner(async_fn, args, to_parent)\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/asyncio_run_in_process/_child.py", line 168, in _run_on_asyncio\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    result: Any = loop.run_until_complete(_do_async_fn(async_fn, args, to_parent, loop))\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/asyncio_run_in_process/_child.py", line 160, in _do_async_fn\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    return await async_fn_task\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/asyncio_run_in_process/_child.py", line 85, in _handle_coro\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    return await coro_task\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/trinity/extensibility/asyncio.py", line 97, in _do_run\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    self.logger.debug("%s: do_run() finished", self)\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/async_generator/_util.py", line 42, in __aexit__\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    await self._agen.asend(None)\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/async_service/asyncio.py", line 458, in background_asyncio_service\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    raise MultiError(\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/async_service/base.py", line 324, in _run_and_manage_task\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    await task.run()\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/async_service/asyncio.py", line 35, in run\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    await self._async_fn(*self._async_fn_args)\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/trinity/extensibility/event_bus.py", line 103, in _auto_connect_new_announced_endpoints\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    await endpoint.connect_to_endpoints(*endpoints_to_connect_to)\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/lahja/asyncio/endpoint.py", line 529, in connect_to_endpoints\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    await asyncio.gather(\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/lahja/asyncio/endpoint.py", line 537, in _await_connect_to_endpoint\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    conn = await AsyncioConnection.connect_to(endpoint.path)\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/home/circleci/repo/.tox/py38-wheel-cli/lib/python3.8/site-packages/lahja/asyncio/endpoint.py", line 88, in connect_to\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    reader, writer = await asyncio.open_unix_connection(str(path))\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "/usr/local/lib/python3.8/asyncio/streams.py", line 111, in open_unix_connection\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'    transport, _ = await loop.create_unix_connection(\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "uvloop/loop.pyx", line 2223, in create_unix_connection\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'  File "uvloop/loop.pyx", line 2218, in uvloop.loop.Loop.create_unix_connection\n'
   DEBUG  11-26 21:59:37  async_process_runner.py  b'BlockingIOError: [Errno 11] Resource temporarily unavailable\n'
```

I've kicked the build.

Elnaril commented:
Thanks for the relaunch of the test, @cburgdorf , it's all green now !! :-)

> Yes, if you could spin out the setup.py fix into its own PR that would be better.

Perfect, I'll do it. Should I also create an issue and reference it in the PR ?

cburgdorf commented:
>Perfect, I'll do it. Should I also create an issue and reference it in the PR ?

Just the PR should be enough

Elnaril commented:
@cburgdorf , I have rebased following the merge of the `argcomplete` version fix into master.
Now, it is the turn of py37-rpc-state-istanbul to fail ! :smile:

gsalgado commented:
> @cburgdorf , I have rebased following the merge of the `argcomplete` version fix into master.
> Now, it is the turn of py37-rpc-state-istanbul to fail ! smile

That's a know issue: #2102

Elnaril commented:
Ah ok, thank you for the info ! :+1: 
So I guess it's worth trying to re-launch this test ? Can someone do it for me, please ? :pray:

gsalgado commented:
I re-ran, but it can fail several times in a row, so better to just ignore it :-(

Elnaril commented:
Indeed, it failed again, so yes it may be better to ignore it then.
Thanks again ! :-)

cburgdorf commented:
Merged! Thanks again @Elnaril for your endurance and time that you invested here!

Elnaril commented:
And thank you @cburgdorf for your time and guidances ! :-)
